### ISSUE TITLE
preg_match compilation failed error

### ISSUE THREAD

**Issue by markvaneijk** [phikiphp/phiki#25]
The error below is shown by most markdown content I tested it on (e.g. https://github.com/rocketeers-app/rocketeers/blob/7137674d7f419bc6146916881f35e600a9a62536/content/articles/zero-downtime-php-deployments.md)

> `preg_match(): Compilation failed: lookbehind assertion is not fixed length at offset 172`

The error does not seem easy to fix because it's a limitation to the PCRE in PHP and the regular expressions are provided by the json files.

ChatGPT explains the error as:

> This PHP error occurs because the regular expression you’re using contains a lookbehind assertion ((?<=...)) that isn’t a fixed length. In regular expressions, lookbehind assertions require the length of the pattern inside to be constant or fixed, meaning each potential match must be of the same length. For example, (?<=abc) is valid because it always looks behind three characters.
>
> However, if you try something like (?<=a*), this is invalid because the * operator allows for a variable length of characters (0 or more as), which makes the lookbehind non-fixed in length.

I fixed it (lazily) using a try/catch: https://github.com/markvaneijk/phiki/commit/5197daff3666aea3f1740ee74c43cbdf974aac12#diff-601d8d0ff3b72b85aa010744f94986847706d35ba6d109c44a3da9e77e177865

**Comment by ryangjchandler** [comment]
Do you know exactly which piece of text was trying to use a variable lookbehind pattern? I might have a way of lowering this syntax into something PCRE-compatible, but if not, I'll have to do some other magic.

**Comment by markvaneijk** [comment]
When I `dd` the regular expression within the try/catch block it shows this one:

```
/(?:(?!^[ \t]*+$)(?:(?<=^until | until |\tuntil |^while | while |\twhile |^elif | elif |\telif |^else | else |\telse |^then | then |\tthen |^do | do |\tdo |^if | if |\tif )|(?<=(?:^|;|\||&|!|\(|\{|\`)))(?:[ \t]*+)(?!nocorrect\W|nocorrect\$|function\W|function\$|foreach\W|foreach\$|repeat\W|repeat\$|logout\W|logout\$|coproc\W|coproc\$|select\W|select\$|while\W|while\$|pushd\W|pushd\$|until\W|until\$|case\W|case\$|done\W|done\$|elif\W|elif\$|else\W|else\$|esac\W|esac\$|popd\W|popd\$|then\W|then\$|time\W|time\$|for\W|for\$|end\W|end\$|fi\W|fi\$|do\W|do\$|in\W|in\$|if\W|if\$))/u
```

**Comment by ryangjchandler** [comment]
@markvaneijk I can't seem to reproduce any exceptions or errors – I did run into a timeout because of a bug in the tokenizer, which I've now fixed and released as `v0.1.2` but would you be able to try again on the new version and report back?

**Comment by markvaneijk** [comment]
I just updated to v0.1.3, but the issue did not resvolve yet.

Tried debugging it a bit and it seems to be in the regular expressions for bash. While testing with the markdown in this file: https://github.com/rocketeers-app/rocketeers/blob/7137674d7f419bc6146916881f35e600a9a62536/content/articles/zero-downtime-php-deployments.md

When I replace all code blocks marked as "bash" to "php" the error is gone and highlighting works. So maybe there's something going on in the bash regexes, the regular expression from the `dd` also seems to be meant for bash.

**Comment by ryangjchandler** [comment]
Interesting.. I've run each of those code blocks through the Shellscript/Bash grammar and didn't get any exceptions or errors.

If you're able to create a standalone reproduction repository, that would be really helpful and I'd really appreciate it, especially since I can't replicate on my machine.

**Comment by markvaneijk** [comment]
As you've tested the code blocks, I suspect it must be something else.. maybe somewhere in how the CommonMark extension works?

I have setup a clean repo with only Laravel + Phiki + spatie/laravel-markdown installed, first commit is the Laravel boilerplate + installed packages and the last three commits to connect everything together.

Repo: https://github.com/markvaneijk/phiki-test

When installed you should be seeing this:

![CleanShot 2024-10-09 at 13 37 47@2x](https://github.com/user-attachments/assets/81036d2b-aeb7-45a6-9da9-ff6f449da5d5)

**Comment by ryangjchandler** [comment]
Thanks @markvaneijk, this is really helpful. Variable-length lookbehinds is something that I'm still trying to resolve, since PCRE doesn't support them, it's actually quite a complex problem.

I've manually patched a few languages that contain them so I'll see what I can do for `shellscript`, but I have a couple of other ideas to try out too.

I've also added a note to the README under a "Known Limitations" section for others who run into the same issue.
